R/lefserPlotClad.R

```diff
@@ -0,0 +1,326 @@
+            name = "Sample", na.value = NA
+        ) +
+        ggplot2::scale_size(name = "Absolute\nscore") +
+        ggtree::theme(legend.position = "right")
+    for (i in collapseThem) {
+        gt2 <- withCallingHandlers(
+            ggtree::collapse(gt2, node = i),
+            warning = function(w) {
+                if (grepl("collapse", w$message)) {
+                    invokeRestart("muffleWarning")
+                }
+            }
+        )
+    }
+    return(gt2)
+}
+
+# Run lefser at all taxonomic levels --------------------------------------
+
+#' Run lefser on all taxonomic levels
+#'
+#' @param relab A SummarizedExperiment.
+#' @param ... Arguments passed to the \code{lefser} function.
+#'
+#' @return An object of class 'lefser_df_all' and 'data.frame'.
+#' @export
+#'
+#' @examples
+#' 
+#' data("zeller14")
+#' z14 <- zeller14[, zeller14$study_condition != "adenoma"]
+#' tn <- get_terminal_nodes(rownames(z14))
+#' z14tn <- z14[tn, ]
+#' z14tn_ra <- relativeAb(z14tn)
+#'
+#' resAll <- lefserAllRanks(relab = z14tn_ra, groupCol = "study_condition")
+#'
+lefserAllRanks <- function(relab,...) {
+    ## Feature names should have the full taxonomy
+    se <- .rowNames2RowData(relab)
+    seL <- mia::splitByRanks(se)
+    ## The kingdom level is not needed
+    ## The mia package doesn't support strain.
+    seL <- seL[names(seL) != "kingdom"]
+    seL <- purrr::map(seL, ~ {
+        seVar <- .x
+        rowDat <- as.data.frame(SummarizedExperiment::rowData(seVar))
+        rowDat <- purrr::discard(rowDat, function(x) all(is.na(x)))
+        rowDat <- S4Vectors::DataFrame(rowDat)
+        SummarizedExperiment::rowData(seVar) <- rowDat
+        seVar
+    })
+    for (i in seq_along(seL)) {
+        rownames(seL[[i]]) <- .lognRowNames(seL[[i]])
+    }
+    
+    res <- seL |>
+        purrr::map(function(x, ...) lefser(relab = x,...), ...) |>
+        dplyr::bind_rows()
+    resOriginal <- lefser(relab, ...)
+    ## Get only tip names (full names with full taxonomy are too long).
+    # resOriginal$features  <- stringr::str_extract(
+    #     resOriginal$features, "[^|]+$"
+    # )
+    res <- res |>
+        ## Avoid repeating features.
+        dplyr::filter(!.data[["features"]] %in% resOriginal$features) |>
+        ## Features not supported by mia are added (strain, OTUs, etc.)
+        dplyr::bind_rows(resOriginal)
+
+    controlVar <- attr(resOriginal, "lgroupf")
+    caseVar <- attr(resOriginal, "case")
+
+    class(res) <- c("lefser_df_all", class(res))
+
+    ## These pathStrings could be used in the plotting function instead (or not)
+    pathStrings <- .selectPathStrings(relab, res)
+    attr(res, "pathStrings") <- pathStrings
+    attr(res, "tree") <- .toTree(pathStrings)
+
+    attr(res, "lgroupf") <- controlVar
+    attr(res, "case") <- caseVar
+    return(res)
+}
+
+## Add taxonomic information to rowData
+## This step is necessary for mia to work
+.rowNames2RowData <- function(x) {
```

Comment:
It seems like this part breaks if the row name of the input `se` is not the concatenation of each clade with the prefix (e.g., gingival dataset which has rowData properly populated already but rowName is in otu).

Reply:
Yeah. That's the case. That's how zelle14 is formatted (taxonomy in rownames). I think we should encourage only one data format (with the taxonomy in the rowData), but we would need either an updated version of zeller14 or use a different dataset.

Reply:
Note that having zeller14 ready already takes to extra steps:
1. Getting terminal nodes.
2. Calculating relative abundance.

A third would be incorporating the taxonomy into the rowData.

Reply:
Also, having the taxonomy in the rowData would imply that all features are of the same level. I think there has been some discussion about it: https://github.com/waldronlab/lefser/issues/41#issuecomment-2171885580

Reply:
Does `lefserAllRanks` require inputs with only the terminal nodes?

Reply:
Ok. What do you think about requiring the input for `lefserAllRanks` function to 1) include only terminal nodes, 2) be relative abundance, and 3) have taxonomy in the rowData? In other words, keep `get_terminal_node`, `relativeAb`, and `.rowNames2RowData` functions outside the `lefserAllRanks` function.

Reply:
> Also, having the taxonomy in the rowData would imply that all features are of the same level. I think there has been some discussion about it: [#41 (comment)](https://github.com/waldronlab/lefser/issues/41#issuecomment-2171885580)

Btw, I don't understand this. Can features be at different levels if the value is `NA`?

Reply:
>Ok. What do you think about requiring the input for lefserAllRanks function to 1) include only terminal nodes, 2) be relative abundance, and 3) have taxonomy in the rowData? In other words, keep get_terminal_node, relativeAb, and .rowNames2RowData functions outside the lefserAllRanks function.

@shbrief, sounds good having the requirement of taxonomy in rowData and having the other functions as helpers to prepare the input. I'll go that route. 

>Btw, I don't understand this. Can features be at different levels if the value is NA?

The terminal nodes are sometimes genus. I think this is only relevant when creating/drawing the cladogram. Probably it won't affect the visualization/interpretation since I'm collapsing clades without any DA taxa.